File: GroceryStore.API/Services/AisleService.cs
using GroceryStore.API.Exceptions.Aisle;
using GroceryStore.API.Models.Requests.Aisle;
using GroceryStore.API.Models.Responses.Aisle;
using GroceryStore.API.Repositories.Interfaces;
using GroceryStore.API.Services.Interfaces;
using System.Threading.Tasks;

namespace GroceryStore.API.Services
{
    public class AisleService : IAisleService
    {
        private readonly IAisleRepository _aisleRepository;

        public AisleService(IAisleRepository repository)
        {
            _aisleRepository = repository;
        }

        public async Task<CreateAisleResponse> CreateAisle(CreateAisleRequest request)
        {
            var resp = await _aisleRepository.GetAisleByName(request.Name);

            if (resp.Aisle?.Name == request.Name)
                throw new DuplicateAisleNameException(request.Name);

            return await _aisleRepository.CreateAisle(request);
        }

        public async Task<GetAisleResponse> GetAisle(string id)
        {
            var resp = await _aisleRepository.GetAisleById(id);

            if (resp.Aisle == null)
                throw new AisleNotFoundException(id);

            return resp;
        }

        public async Task<GetAislesResponse> GetAisles()
        {
            return await _aisleRepository.GetAisles();
        }

        public async Task<UpdateAisleResponse> UpdateAisle(UpdateAisleRequest request)
        {
            var resp = await _aisleRepository.GetAisleById(request.Id);

            if (resp.Aisle == null)
                throw new AisleNotFoundException(request.Id);

            var nameResp = await _aisleRepository.GetAisleByName(request.Name);

            if (resp.Aisle.Name != request.Name && nameResp.Aisle != null)
                throw new DuplicateAisleNameException(request.Name);

            return await _aisleRepository.UpdateAisle(request);
        }

        public async Task<DeleteAisleResponse> DeleteAisle(string id)
        {
            var resp = await _aisleRepository.GetAisleById(id);

            if (resp.Aisle == null)
                throw new AisleNotFoundException(id);

            return await _aisleRepository.DeleteAisle(id);
        }
    }
}

File: GroceryStore.Web/src/app/services/aisle.service.ts
import { Injectable, signal } from '@angular/core';
import { IAisleService } from '../interfaces/aisle-service.interface';
import { map, Observable, tap } from 'rxjs';
import { CreateAisleRequest, UpdateAisleRequest, DeleteAisleRequest } from '../models/requests/aisle-requests';
import { GetAislesResponse, GetAisleResponse, UpdateAisleResponse, DeleteAisleResponse } from '../models/responses/aisle-responses';
import { environment } from '../../environments/environment';
import { HttpClient } from '@angular/common/http';
import { Aisle } from '../models/aisle';

@Injectable({
  providedIn: 'root'
})
export class AisleService implements IAisleService {
    private apiUrl = `${environment.apiUrl}/aisles`

    private aisles = signal<Aisle[]>([]);

    constructor(private http: HttpClient) { }

    getAisles(): Observable<GetAislesResponse> {
        return this.http.get<GetAislesResponse>(`${this.apiUrl}`).pipe(
                tap((response) => {
                this.aisles.set(response.aisles);
            }));
    }

    getAisleById(id: string): Observable<GetAisleResponse> {
        return this.http.get<GetAisleResponse>(`${this.apiUrl}/${id}`);
    }

    createAisle(request: CreateAisleRequest): Observable<GetAisleResponse> {
        return this.http.post<GetAisleResponse>(`${this.apiUrl}/create`, request);
    }

    updateAisle(request: UpdateAisleRequest): Observable<UpdateAisleResponse> {

        return this.http.put<UpdateAisleRequest>(`${this.apiUrl}/update`, request);
    }

    deleteAisle(id: string): Observable<DeleteAisleResponse> {
        return this.http.delete<DeleteAisleRequest>(`${this.apiUrl}/delete/${id}`); 
    }

}
